store/postgres/migrations/2019-03-11-221621_delete_unneeded_entity_history/up.sql

```diff
@@ -0,0 +1,190 @@
+    -- Get operation type and source
+    IF (TG_OP = 'INSERT') THEN
+        operation_type := 0;
+        event_source := NEW.event_source;
+        subgraph := NEW.subgraph;
+        entity := NEW.entity;
+        entity_id := NEW.id;
+        data_before := NULL;
+    ELSIF (TG_OP = 'UPDATE') THEN
+        operation_type := 1;
+        event_source := NEW.event_source;
+        subgraph := OLD.subgraph;
+        entity := OLD.entity;
+        entity_id := OLD.id;
+        data_before := OLD.data;
+    ELSIF (TG_OP = 'DELETE') THEN
+        operation_type := 2;
+        event_source := current_setting('vars.current_event_source', TRUE);
+        subgraph := OLD.subgraph;
+        entity := OLD.entity;
+        entity_id := OLD.id;
+        data_before := OLD.data;
+    ELSE
+        RAISE EXCEPTION 'unexpected entity row operation type, %', TG_OP;
+    END IF;
+
+    IF event_source = 'REVERSION' THEN
+        is_reversion := TRUE;
+    END IF;
+
+    SELECT id INTO event_id
+    FROM event_meta_data
+    WHERE db_transaction_id = txid_current();
+
+    new_event_id := null;
+
+    IF event_id IS NULL THEN
+        -- Log information on the postgres transaction for later use in revert operations
+        INSERT INTO event_meta_data
+            (db_transaction_id, db_transaction_time, source)
+        VALUES
+            (txid_current(), statement_timestamp(), event_source)
+        RETURNING event_meta_data.id INTO new_event_id;
+    END IF;
+
+    -- Log row metadata and changes, specify whether event was an original ethereum event or a reversion
+    INSERT INTO entity_history
+        (event_id, entity_id, subgraph, entity, data_before, reversion, op_id)
+    VALUES
+        (COALESCE(new_event_id, event_id), entity_id, subgraph, entity, data_before, is_reversion, operation_type);
+    RETURN NULL;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Remove unneeded entity_history data, and event_meta_data that becomes
+-- unnecessary because of that, too. Since we need to get rid of most of
+-- the data, it is faster to copy the 'good' data into a new table and swap
+-- tables around
+
+-- Create a new table that clones the entity_history table. For now,
+-- we only clone the schema. We create indexes and constraints after
+-- we are done copying, for speed.
+create table eh(like entity_history
+                     including defaults
+                     including comments
+                     including storage);
+-- eh and entity_history both use entity_history_id_seq now. Remove its
+-- use from entity_history, so we don't get a conflict later when we
+-- drop that table
+alter table entity_history
+      alter column id drop default;
+alter sequence entity_history_id_seq
+      owned by eh.id;
+alter table eh drop column data_after;
+
+-- Lock any writes to entity_history and event_meta_data. We can not
+-- afford to have any changes happen while we copy, as we would lose
+-- them because we are copying
+lock table entity_history in exclusive mode;
```

Comment:
Is this something we should do _before_ dropping the `id` column? Just curious if there could be any race conditions here.

Reply:
I added the sequence moving stuff after the locking, and so this should move up, even though dropping the default for the `id` column will take an `access exclusive` lock, which is even stricter than the exclusive lock, and also prevents any queries against `entity_history` from happening. I need to check though that we are not going to shoot ourselves in the foot with these two locks on the same table.